File: internal/extensions/plugin/contracts.go
package plugin

type Capability string

const (
	CapabilityUICommands Capability = "ui.commands"
	CapabilityUIMenu     Capability = "ui.menu"
	CapabilityDataSource Capability = "data.providers"
)

type Manifest struct {
	ID            string       `json:"id"`
	Version       string       `json:"version"`
	MinAppVersion string       `json:"min_app_version"`
	Capabilities  []Capability `json:"capabilities"`
}

type CommandContribution struct {
	ID         string `json:"id"`
	Title      string `json:"title"`
	HandlerKey string `json:"handler_key"`
}

type DataProviderContribution struct {
	ID           string   `json:"id"`
	ResourceType []string `json:"resource_types"`
	QueryHook    string   `json:"query_hook"`
}

type Contribution struct {
	Manifest      Manifest                   `json:"manifest"`
	Commands      []CommandContribution      `json:"commands,omitempty"`
	DataProviders []DataProviderContribution `json:"data_providers,omitempty"`
	Metadata      map[string]string          `json:"metadata,omitempty"`
}

File: frontend/src/features/plugin/contracts.ts
export type PluginCapability =
    | 'ui.commands'
    | 'ui.menu'
    | 'data.providers';

export type PluginCommandCategory = 'Editor' | 'Layout' | 'Connection' | 'View' | 'App';

export interface PluginMenuItem {
    id: string;
    label: string;
    shortcut?: string;
    disabled?: boolean;
    danger?: boolean;
    hasBadge?: boolean;
    action: () => void | Promise<void>;
}

export interface PluginManifest {
    id: string;
    version: string;
    minAppVersion: string;
    capabilities: PluginCapability[];
}

export interface PluginCommandContribution {
    id: `ext.${string}`;
    title: string;
    category: PluginCommandCategory;
    defaultBinding: string;
    defaultWhen?: string;
    handlerKey: string;
}

export interface PluginDataProviderContribution {
    id: string;
    resourceTypes: string[];
    queryHook: string;
}

export interface PluginMenuContribution {
    sectionId: string;
    sectionTitle?: string;
    item: PluginMenuItem;
}

export interface PluginContribution {
    manifest: PluginManifest;
    commands?: PluginCommandContribution[];
    dataProviders?: PluginDataProviderContribution[];
    menus?: PluginMenuContribution[];
}

export interface ResolvedPluginCommand {
    pluginId: string;
    entry: {
        id: `ext.${string}`;
        label: string;
        category: PluginCommandCategory;
        defaultBinding: string;
        defaultWhen?: string;
        action: () => void | Promise<void>;
    };
}
